### PR TITLE
Allow userland middlewares

### PR DESCRIPTION
**Pull request by l-vo**
Related to the needs to replace SQLLogger by Middlewares, this implementations allows to easily inject middleware via a `doctrine.middleware` tag.

### CODE REVIEW

**Comment by greg0ire** [comment]
You can run `vendor/bin/phpcbf` to fix most CS issues :)

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (greg0ire) [inline DependencyInjection/Compiler/MiddlewaresPass.php]:
```suggestion
final class MiddlewaresPass implements CompilerPassInterface
```

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (greg0ire) [inline DependencyInjection/Compiler/MiddlewaresPass.php]:
```suggestion
    private $connectionDefsParam;
```
:fr: spotted :slightly_smiling_face:

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (greg0ire) [inline DependencyInjection/Compiler/MiddlewaresPass.php]:
```suggestion
    public function process(ContainerBuilder $container): void
```

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (greg0ire) [inline DependencyInjection/Compiler/MiddlewaresPass.php]:
```suggestion
        foreach ( $container->getParameter($this->connectionDefsParam) as $name => $id) {
```

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (greg0ire) [inline DependencyInjection/Compiler/MiddlewaresPass.php]:
```suggestion
            $middlewareAbstractDefs[$id] = $container->getDefinition($id);
```

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (greg0ire) [inline DependencyInjection/Compiler/MiddlewaresPass.php:50]: I get the need for early continue/return, etc. but here it seems needlessly convoluted as it requires a logical inversion and an extra statement.

```suggestion
                if (is_subclass_of($abstractDef->getClass(), ConnectionNameAwareInterface::class)) {
                    $childDef->addMethodCall('setConnectionName', [$name]);
                }
```

File: Middleware/ConnectionNameAwareInterface.php
Comment (greg0ire) [inline Middleware/ConnectionNameAwareInterface.php:8]: What's the use case for this?

**Comment by l-vo** [comment]
> You can run `vendor/bin/phpcbf` to fix most CS issues :)

Thank you, I missed that cs failed

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (l-vo) [reply DependencyInjection/Compiler/MiddlewaresPass.php]: 😅 thank you

File: Middleware/ConnectionNameAwareInterface.php
Comment (l-vo) [reply Middleware/ConnectionNameAwareInterface.php:8]: I wrote this PR with the replacement of `DebugStack` by middlewares in mind. In the symfony profiler, queries are grouped by connection. I faced a choice:
- either I inject in the middleware a different service (kind of debug data holder) by connection
- either the middleware is aware of the connection and set queries in an unique debug data holder tied with their connection

It's not a big deal to duplicate an abstract middleware for each connection with a tag, but duplicating too the debug data holder seemed to me more complicated; it's why I chose the second possibility.

This is how I plan to create the debug data holder:
```php
class DebugDataHolder
{
    private array $data = [];

    public function addQuery(string $connexionName, Query $query): void
    {
        $this->data[$connexionName][] = [
            'sql' => $query->getSql(),
            'params' => $query->getParams(),
            'types' => $query->getTypes(),
            'executionMS' => static fn() => $query->getDuration(),  // stop() may not be called at this point
        ];
    }

    public function getData(string $connexionName): array
    {
        if (!isset($this->data[$connexionName])) {
            return [];
        }

        foreach ($this->data[$connexionName] as &$data) {
            if (is_callable($data['executionMS'])) {
                $data['executionMS'] = $data['executionMS']();
            }
        }

        return $this->data[$connexionName];
    }
}
```

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (ostrolucky) [inline DependencyInjection/Compiler/MiddlewaresPass.php]: Why would this be configurable here if we hardcode it in DoctrineExtension?

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [inline DependencyInjection/DoctrineExtension.php]: Why was this moved out of `loadMiddlewares`?

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [inline DependencyInjection/DoctrineExtension.php]: This disables logging even when it's enabled, right? Basically it will block release of 2.6 once we merge this, until you finish follow-up PR

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (l-vo) [reply DependencyInjection/Compiler/MiddlewaresPass.php]: Not needed here indeed

File: DependencyInjection/DoctrineExtension.php
Comment (l-vo) [reply DependencyInjection/DoctrineExtension.php]: It disables logging but the logging is now done by the logging middleware, see `Resources/config/middlewares.xml` in this PR.

File: DependencyInjection/DoctrineExtension.php
Comment (l-vo) [reply DependencyInjection/DoctrineExtension.php]: I moved nothing out of `loadMiddlewares`, `loadMiddleware` has just been added in this PR. What am I missing ?

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [reply DependencyInjection/DoctrineExtension.php]: that's just abstract middleware

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [reply DependencyInjection/DoctrineExtension.php]: useMiddlewaresIfAvailable was renamed to loadMiddlewares. This condition was there.

Comment (l-vo) [reply DependencyInjection/DoctrineExtension.php]: Ok sorry, it's a subjective choice, in my mind, `loadMiddleware` should only register middlewares, not disable what is replaced by a middleware. But I can move it back in `loadMiddleware` if you think it's better.

File: DependencyInjection/DoctrineExtension.php
Comment (l-vo) [reply DependencyInjection/DoctrineExtension.php]: Yes but abstract middlewares are registered for each connection by `MiddlewarePass` 🙂

Comment (ostrolucky) [reply DependencyInjection/DoctrineExtension.php]: I see. And which part disables middleware when $config['logging'] is false?

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [reply DependencyInjection/DoctrineExtension.php]: Point is to guarantee logger argument is replaced only when middlewares are registered. After this split, nothing guarantees it anymore and one can call these separately. Perhaps old name was better then, since `useMiddlewares` doesn't imply it only loads them.

Comment (l-vo) [reply DependencyInjection/DoctrineExtension.php]: The tests guarantee that 🙂 But I understand the need from a code point of view, I moved it back into `useMiddlewaresIfAvailable` (I renamed the method too).

File: DependencyInjection/DoctrineExtension.php
Comment (l-vo) [reply DependencyInjection/DoctrineExtension.php]: You're right, I missed this case, to allow to handle this case, I did some other changes:
- Middleware can now be registered only for chosen connections (with a tag attribute)
- An attribute `AsMiddleware` has been added

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (ostrolucky) [inline DependencyInjection/Compiler/MiddlewaresPass.php]:
```suggestion
                if (isset($middlewareConnections[$id]) && ! in_array($name, $middlewareConnections[$id], true)) {
```

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [inline DependencyInjection/DoctrineExtension.php]: small optimization: `$container->getDefinition('doctrine.dbal.logging_middleware')` should be assigned to variable outside of loop so DI doesn't have to search for it on each iteration

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (ostrolucky) [inline Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]: in future instead please use syntax like such `@return array<string, array{0: class-string, 1: bool}>`

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (ostrolucky) [inline Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]:
```suggestion
    class AutoconfiguredPHP7Middleware implements Middleware
```

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (ostrolucky) [inline Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]:
```suggestion
        class AutoconfiguredMiddleware
```

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (ostrolucky) [inline Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]:
```suggestion
        class AutoconfiguredConnectionAwareMiddleware
```

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (ostrolucky) [inline Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]:
```suggestion
class ConnectionAwarePHP7Middleware implements ConnectionNameAwareInterface
```

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (ostrolucky) [inline Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]:
```suggestion
class PHP7Middleware
```

File: DependencyInjection/DoctrineExtension.php
Comment (ostrolucky) [inline DependencyInjection/DoctrineExtension.php]: early return would be better to avoid nesting

**Review by ostrolucky — CHANGES_REQUESTED** [review]
Nice work, I think there are only minor comments

File: DependencyInjection/Compiler/MiddlewaresPass.php
Comment (l-vo) [reply DependencyInjection/Compiler/MiddlewaresPass.php]: Indeed, I don't know why I used array_search 😅

File: Tests/DependencyInjection/Compiler/MiddlewarePassTest.php
Comment (l-vo) [reply Tests/DependencyInjection/Compiler/MiddlewarePassTest.php]: I took the freedom to rename it as `AutoconfiguredMiddlewareWithConnection` instead to avoid confusion with `ConnectionNameAwareInterface` (a middleware can target a specific connection without being aware of this connection name); is it ok for you ?

**Comment by l-vo** [comment]
@ostrolucky thank you, all comments are addressed.

**Review by ostrolucky — APPROVED** [review]

**Comment by ostrolucky** [comment]
LGTM, but would be nice if community properly tested this. I didn't try it in a project

**Comment by greg0ire** [comment]
@l-vo sorry to comment about this more than 1 year later, is there documentation on this somewhere?

**Comment by l-vo** [comment]
@greg0ire yes you're right, I forgot to document the configuration. I'm going to do that soon.

**Comment by greg0ire** [comment]
Great! Thanks in advance!

**Comment by l-vo** [comment]
@greg0ire done in #1653 :)

**Pull request merged**
